Modal for adding topics to classroom

DorianBD opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *all* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes part of #18470 .
2. This PR does the following: Enhance the user experience on the classroom admin page. The goal is to improve the process of adding a topic by presenting the user with a list of unused topics that can be added to the classroom with a simple click, eliminating the need for the user to manually input topic IDs, which is inconvenient.
This issue is resolved in two pull requests (this is the second one):

Create a new modal component called "add-topic-to-classroom-modal.component". This modal component will be opened when the user clicks on "add a topic," in line with the issue description.
This modal component will display all topics that are currently not associated with any classrooms as checkboxes. When the user clicks "Add Topics" the modal will close and return to the parent component (classroom-admin-page.component) a list of topic IDs that the user wishes to add. The parent component will then call the existing "addTopicId()" function to add each topic to the classroom using its ID.

I also had to create a function that call the backend handler for unused topics in the editable-topic-backend-api.service.ts file.

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...", followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [ ] I have assigned the correct reviewers to this PR (or will leave a comment with the phrase "@{{reviewer_username}} PTAL" if I don't have permissions to assign reviewers directly).
- [x] The linter/Karma presubmit checks pass on my local machine, and my PR follows the [coding style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide)).
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)

## Proof that changes are correct

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface. Please also include
videos/screenshots of the developer tools browser console, so that we can be
sure that there are no console errors.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->


https://github.com/oppia/oppia/assets/144214967/6817180b-4b50-4821-9ac2-5109fc92c228

DorianBD committed:
add getUnusedTopicsAsync to the editable-topic-backend-api.service.ts Ref #18470

DorianBD committed:
correct typo in the url handler name Ref #18470

DorianBD committed:
feat: create add-topics-to-classroom modal Ref #18470

DorianBD committed:
test: adding test for add-topic-to-classroom modal Ref #18470

DorianBD committed:
test:  add test for openAddTopicModal() Ref #18470

DorianBD committed:
refactor: test of openAddTopicModal() Ref #18470

DorianBD committed:
refactor:  add getUnusedTopicsAsync fails test and correct error of method getUnusedTopicsAsync Ref #18470

DorianBD committed:
feat: add scroll and loading message Ref #18470

DorianBD committed:
correct test error due to not declared component Ref #18470

DorianBD committed:
test: add tests for add-topic-to-classroom-modal to cover all lines Ref #18470

DorianBD committed:
correct linter errors Ref #18470

DorianBD committed:
refactor: using fakeAsync and tick instead of async for add-topic-to-classroom-modal test Ref #18470

DorianBD committed:
correct linter errors Ref #18470

DorianBD committed:
correct linter errors Ref #18470

DorianBD committed:
correct linter errors Ref #18470

DorianBD committed:
correct linter errors Ref #18470

DorianBD committed:
correct linter errors test remove antislash Ref #18470

DorianBD committed:
correct linter errors Ref #18470

DorianBD committed:
correct typescripts errors Ref #18470

DorianBD committed:
correct linters errors Ref #18470

DorianBD committed:
correct linters errors Ref #18470

DorianBD committed:
correct typescript error: UnusedTopicsResponse was not yet exported Ref #18470

DorianBD committed:
correct typescript error: provide all the properties of Topic Backend dict Ref #18470

DorianBD committed:
correct lint errors Ref #18470

kevintab95 reviewed (changes requested):
Hi @DorianBD, could you please upload a video for the "Proof that changes are correct" section -- it currently shows some text content instead?

DorianBD commented:
@kevintab95 PTAL for review. Yes sorry apparently I didn't uploaded  the video properly.

kevintab95 reviewed (approved):
LGTM, thanks @DorianBD!

seanlip committed:
Merge branch 'develop' into modal-for-adding-topics-to-classroom

seanlip commented:
@DorianBD The e2e tests are failing on this PR, and I think the failure is related to your changes since it relates to the classroom page. Could you PTAL? Thanks.

DorianBD commented:
Hi @seanlip, yeah it seems to be related to my changes. I am going to take a look. Thanks

DorianBD commented:
Hi @seanlip, can you PTAL. The error is as follows: "Error: Add topic ID to classroom input is not visible." in the test "/core/tests/webdriverio_desktop/classroomPageFileUploadFeatures.js". I think this test calls a part of the code that I deleted because I refactored the logic. But I'm not quite understanding how this part of the code base works. Can you give me a hint on where the error might be coming from, thank you.

seanlip commented:
@DorianBD We have a guide on e2e tests here: https://github.com/oppia/oppia/wiki/End-to-End-Tests. I would start by running the test locally on your computer and watching it run in the browser, it should make it obvious where the problem is. Can you try doing that?

If you still have issues, please capture the necessary info in a [debugging doc](https://github.com/oppia/oppia/wiki/Debugging-Docs) so that we can better help you. Thanks!

seanlip commented:
Hi @DorianBD -- just following up, any progress on this? Thanks!

DorianBD commented:
Hi @seanlip, I had hardware issues and had to get another PC. I reinstalled the project, but when I run the end-to-end tests with "python -m scripts.run_e2e_tests --suite="classroomPageFileUploadFeatures"," I encounter this error:
![image](https://github.com/oppia/oppia/assets/144214967/e4ecab22-5898-44e5-bb09-f387e79e58be)

seanlip commented:
@DorianBD Thanks for the update -- hm, I'm not sure why this is happening, though I suspect it's due to your PC. Have you tried searching for the error message and seeing what others have done?

DorianBD commented:
> @DorianBD Thanks for the update -- hm, I'm not sure why this is happening, though I suspect it's due to your PC. Have you tried searching for the error message and seeing what others have done?

Hi @seanlip, I tried reinstalling ubuntu and the project again but I still have the same issue. I didn't find discussions related to this error in the github discussions section.

seanlip commented:
@DorianBD I don't mean on GitHub Discussions, I mean searching on the web. Look for something like "sock bind oserror 95 Operation not supported", and take a look at the code in run_portserver.py. Can you find anything?

Ash-2k3 commented:
Hi @DorianBD, just following up on PRs with "LGTM" label, any progress on this ? Thanks.

seanlip commented:
Closing this PR since there is no activity. Will update the issue thread accordingly.
